edward/inferences/replica_exchange_mc.py

```diff
@@ -0,0 +1,284 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import six
+import tensorflow as tf
+import numpy as np
+
+from collections import OrderedDict
+from edward.inferences.monte_carlo import MonteCarlo
+from edward.models import Empirical, RandomVariable
+from edward.util import check_latent_vars, copy
+
+
+class memory_lambda:
+  """
+  Class to use instead of lambda.
+  lambda is affected by the change of x,
+  so memory_lambdda output x at the time of definition.
+  """
+
+  def __init__(self, x):
+    self.x = x
+
+  def __call__(self):
+    return self.x
+
+
+class ReplicaExchangeMC(MonteCarlo):
+  """
+  Replica Exchange MCMC(https://en.wikipedia.org/wiki/Parallel_tempering)
+
+  #### Examples
+  ```python
+  cat = Categorical(probs=[0.5,0.5])
+  x = Mixture(cat=cat, components=
+                [MultivariateNormalDiag([0.0,0.0], [1.0,1.0]),
+                 MultivariateNormalDiag([10.0,10.0], [1.0,1.0])])
+  proposal_x = MultivariateNormalDiag(x, [1.0,1.0])
+  qx = Empirical(tf.Variable(tf.zeros([10000, 2])))
+  inference = ed.ReplicaExchangeMC(latent_vars={x: qx},
+                                 proposal_vars={x: proposal_x})
+  ```
+  """
+
+  def __init__(self, latent_vars, proposal_vars, data=None,
+               betas=np.logspace(0, -2, 5), exchange_freq=0.1):
+    """Create an inference algorithm.
+    Args:
+      proposal_vars: dict of RandomVariable to RandomVariable.
+        Collection of random variables to perform inference on; each is
+        binded to a proposal distribution $g(z' \mid z)$.
+      betas: list of inverse temperature.
+      exchange_freq: frequency of exchanging replica
+    """
+    check_latent_vars(proposal_vars)
+    self.proposal_vars = proposal_vars
+
+    self.n_replica = len(betas)
+    if betas[0] != 1:
+      raise ValueError("betas[0] must be 1.")
+    self.betas = [tf.constant(beta, dtype=list(latent_vars.values())[0].dtype)
```

Comment:
Use `tf.convert_to_tensor` instead of `tf.constant`.

Reply:
OK. I'll fix.